x/mongo/driver/operation.go

```diff
@@ -1459,7 +1459,65 @@ func (op Operation) addReadConcern(dst []byte, desc description.SelectedServer)
 	return bsoncore.AppendDocumentElement(dst, "readConcern", data), nil
 }
 
-func (op Operation) addWriteConcern(dst []byte, desc description.SelectedServer) ([]byte, error) {
+func getWriteConcernTimeout(ctx context.Context, op Operation) time.Duration {
+	if contextDeadline, ok := ctx.Deadline(); ok {
+		return time.Until(contextDeadline)
+	}
+
+	if op.Timeout != nil {
+		return *op.Timeout
+	}
+
+	if op.Client == nil {
+		return 0
+	}
+
+	return op.Client.CurrentWTimeout
+}
+
+func marshalBSONWriteConcern(wc writeconcern.WriteConcern, wtimeout time.Duration) (bsontype.Type, []byte, error) {
```

Comment:
Move the writeconcern marshaling logic to the operation level to avoid exporting it in writecocnern. The writeconcern version will be removed in GODRIVER-2367